LD2420 Documentation (#2956)
OttoWinter/esphomedocs: components/sensor/ld2420.rst

      stop_bits: 1

    # The LD2420 has 16 time gate intervals (0-15) with a gate
    # resolution of 0.75 meters
    ld2420:
      presence_time_window: 120s
      detection_gate_min: 1
      detection_gate_max: 12
      g0_move_threshold: 60000
      g0_still_threshold: 40000
      g1_move_threshold: 30000
      g1_still_threshold: 20000
      g2_move_threshold: 400
      g2_still_threshold: 200
      g3_move_threshold: 250
      g3_still_threshold: 200
      g4_move_threshold: 250
      g4_still_threshold: 200
      g5_move_threshold: 250
      g5_still_threshold: 200
      g6_move_threshold: 250
      g6_still_threshold: 200
      g7_move_threshold: 250
      g7_still_threshold: 150
      g8_move_threshold: 250
      g8_still_threshold: 150
      g9_move_threshold: 250
      g9_still_threshold: 100
      g10_move_threshold: 250
      g10_still_threshold: 100
      g11_move_threshold: 250
      g11_still_threshold: 100
      g12_move_threshold: 250
      g12_still_threshold: 100
      g13_move_threshold: 200
      g13_still_threshold: 100
      g14_move_threshold: 200
      g14_still_threshold: 100
      g15_move_threshold: 200
      g15_still_threshold: 100

> .. note::
> 
>     For UART configuration, ``baud_rate``, ``parity`` and ``stop_bits`` **must be** respectively ``115200 or 256000``, ``NONE`` and ``1``.
>     Use of hardware UART pins is critical in order to support the out-of-the-box 256000 or 115200 baud rates of the LD2420 sensor.

Reviewer: Given the issues I see already mentioned on the PR, it might be better to make this a `warning` and move the bit about the hardware UART to the top, like this:

```suggestion
.. warning::

    Use of a hardware UART is critical in order to support the out-of-the-box 256000 or 115200 baud rates of the LD2420 sensor. Communication problems are likely to result otherwise.
    In addition, for UART configuration, ``baud_rate``, ``parity`` and ``stop_bits`` **must be** respectively ``115200 or 256000``, ``NONE`` and ``1``.
```

Configuration variables:
************************

The configuration is made up of three parts: The LD2420 component, individual sensors,
and binary sensors. These may change, this is a early component release based on firmware v1.5.3.
*There are clearly undocumented functions in the firmware which are not available at this time.*

> - **presence_time_window** (*Optional*, int): Time in seconds during which presence state will stay present
>   after leaving. Any move or still energy detection within the time window restarts a count down from the set value.

Reviewer: ```suggestion
- **presence_time_window** (*Optional*, int): The time in seconds during which the occupied state (presence) will persist
  after presence is no longer detected. Any energy detection within the time window restarts the countdown from this value.
```

  Defaults to ``120s``.
- **detection_gate_min** (*Optional*, int): Minimum distance for move or still energy detection.
  Value between gate ``1`` and ``detection_gate_max - 1`` incrementally (each increment equals ``0.75m`` at this time).

>   Defaults to ``1``.

Reviewer: ```suggestion
- **detection_gate_min** (*Optional*, int): Minimum distance for move or still energy detection.
  Value between 0 and ``detection_gate_max - 1``. Each increment equals 70 cm.
  Defaults to 1.
```

- **detection_gate_max** (*Optional*, int): Maximum gate for movement detection.
  Value between ``1`` to ``15`` each gate detects movement and still energy within ``0.75m`` of range incrementally.
  Defaults to ``12``.
  With maximum distance equates to ``16`` * ``0.75m`` resulting in a possible ``12m`` however accuracy at ``12m`` is not certain.

>   The recommended maximum value is ``12`` * ``0.75`` which results in approximately ``9m``.

Reviewer: ```suggestion
  Value from 1 to 15. Each gate detects movement and still energy at 70 cm increments.
  Defaults to 12.
  At the maximum value, the sensor's range equates to 16 * 70 cm resulting in an
  approximately 12-meter range; however, accuracy at this range may be diminished.
  The recommended maximum value is 12 * 70 cm which results in approximately 9 meters.
```

Also, it's really not necessary to monospace every number. haha 😇

- **g#_move_threshold** (*Optional*, int): Gate threashold level for motion energy detection (Gate # from 0 to 15).

>   Above this level for the considered gate (distance), movement detection will be triggered.

Reviewer: ```suggestion
  A value greater than that specified for the gate (distance) will trigger movement detection.
```

>   Defaults to ``see table below``.

Reviewer: Make this a local anchor/link

Author: I don't understand how to local anchor/link to a table.

Reviewer: Fixed 😄

Reviewer: See new comment below

- **g#_still_threshold** (*Optional*, int): Threshold for the Xth gate for still detection. (Gate # from 0 to 15).

>   Above this level for the considered gate (distance), still detection will be triggered.

Reviewer: ```suggestion
- **g#_still_threshold** (*Optional*, int): Gate threashold level for still energy detection (Gate # from 0 to 15).
  A value less than that specified for the gate (distance) will trigger still detection.
```

>   Defaults to ``see table below``.

Reviewer: Make this a local anchor/link

.. list-table:: Default values for gate threshold
    :widths: 25 25 25
    :header-rows: 1

    * - Gate:
      - Default: Move threshold
      - Default: Still threshold
    * - 0
      - 60000
      - 40000
    * - 1
      - 30000
      - 20000
    * - 2
      - 400
      - 200
    * - 3
      - 300
      - 250
    * - 4
      - 250
      - 150
    * - 5
      - 250
      - 150
    * - 6
      - 250
      - 150
    * - 7
      - 250
      - 150
    * - 8
      - 300
      - 150
    * - 9
      - 250
      - 150
    * - 10
      - 250
      - 150
    * - 11
      - 250
      - 150
    * - 12
      - 250
      - 100
    * - 13
      - 200
      - 100
    * - 14
      - 200
      - 100
    * - 15
      - 200
      - 100

Sensor
------

> The ``ld2420`` sensor allows you to use your :doc:`ld2420` to sense mmWave distance measurements.

Reviewer: ```suggestion
The ``ld2420`` sensor allows you to use your :doc:`ld2420` to approximate the distance
between the sensor and an object moving within its field of detection.
```

Reviewer: I think this might've gotten paved when you reworked the docs post-refactoring. This is just a suggestion; I just think the way it is phrased here is quite awkward.

.. code-block:: yaml

    sensor:
      - platform: ld2420
        moving_distance:
          name : Moving Distance

Configuration variables:
************************

> - **moving_distance** (*Optional*): Distance of detected moving target.

Reviewer: ```suggestion
- **moving_distance** (*Optional*): Distance between the sensor and the detected moving target.
```

  All options from :ref:`Sensor <config-sensor>`.

Binary Sensor
-------------

> The ``ld2420`` binary sensor allows you to use your :doc:`ld2420` to sence presence.

Reviewer: ```suggestion
The ``ld2420`` binary sensor allows you to use your :doc:`ld2420` to sense presence.
```

.. code-block:: yaml

    binary_sensor:
      - platform: ld2420
        has_target:
          name: Presence

Configuration variables:
************************

- **has_target** (*Optional*): If true target detect either still or in movement.

>   All options from :ref:`Binary Sensor <config-binary_sensor>`.

Reviewer: There is some consistency issue here...it says "if true" but the example doesn't illustrate that...maybe explain what's going on here?

Author: Done.

Important Information:
**********************

> Solid objects and noise outside the ``detection_gate_max`` and ``detection_gate_min`` can cause false
> detections or abnormal gate thresholds. For example if the gate max range includes a wall it can potentially
> generate signal reflections. If you see unusual detections confirm the issue by placing the sensor in a
> completely open area. Never place two sensors in each others detection field, this will certainly
> cause false detections. Noise is a problem in many enviroments and future releases of this component
> will have an auto noise floor configuration option once Hi-Link provided full documentation.

Reviewer: ```suggestion
Solid objects and noise outside the ``detection_gate_max`` and ``detection_gate_min`` may cause false
detections or result in abnormal gate thresholds. For example, a wall within the gate max range can result
in signal reflections. If your sensor reports unexpected detections, you should test it by placing it in a
completely open room with no moving objects.

Also, never place two or more sensors in a manner such that their detection fields overlap, as this will certainly
result in false detections.

Noise is a problem in many environments and we anticipate that future releases of this component will have an auto
noise floor configuration option.
```

- **presence_time_window** (*Optional*, int): The time in seconds during which the occupied state (presence) will persist
  after presence is no longer detected. Any energy detection within the time window restarts the countdown from this value.
  Defaults to ``120s``.
- **detection_gate_min** (*Optional*, int): Minimum distance for move or still energy detection.
  Value between 0 and ``detection_gate_max - 1``. Each increment equals 70 cm.
  Defaults to 1.
- **detection_gate_max** (*Optional*, int): Maximum gate for movement detection.
  Value from 1 to 15. Each gate detects movement and still energy at 70 cm increments.
  Defaults to 12.
  At the maximum value, the sensor's range equates to 16 * 70 cm resulting in an
  approximately 12-meter range; however, accuracy at this range may be diminished.
  The recommended maximum value is 12 * 70 cm which results in approximately 9 meters.
- **g#_move_threshold** (*Optional*, int): Gate threashold level for motion energy detection (Gate # from 0 to 15).
  A value greater than that specified for the gate (distance) will trigger movement detection.

>   Defaults to ``see table below``.
> - **g#_still_threshold** (*Optional*, int): Gate threashold level for still energy detection (Gate # from 0 to 15).
>   A value less than that specified for the gate (distance) will trigger still detection.
>   Defaults to ``see table below``.
> 
> .. list-table:: Default values for gate threshold

Reviewer: This should do it:

```suggestion
  Defaults are as shown in :ref:`the table below <ld2420-default_values_for_gate_threshold>`.
- **g#_still_threshold** (*Optional*, int): Gate threashold level for still energy detection (Gate # from 0 to 15).
  A value less than that specified for the gate (distance) will trigger still detection.
  Defaults are as shown in :ref:`the table below <ld2420-default_values_for_gate_threshold>`.

.. _ld2420-default_values_for_gate_threshold:

.. list-table:: Default Values for Gate Threshold
```

Author: Tests good, updating it.

  The report is provided in the log console and for information only. In order to calibrate the gate configurations no
  objects should be present in the target environment. The recording will establish the mmWave noise floor levels and
  undesirable reflections or signal distortions peaks. This should be collected for at least 30 seconds. Longer runs can
  catch more potential false signals when devices near the module are used such as Microwave ovens and other signal
  emiting sources.
  When the **apply_config** button is triggered in calibration mode the ld2420 will be configured with calculated values
  based on its collected noise floor data and the module will be returned to Normal Mode. Selecting any other mode drops
  the sessions collected noise data and no changes setting are staged. Once the **apply_config** action is performed it
  will persist. The gate values can be manually adjusted or calibrated again or you can trigger the
  factory reset button.
- **Simple** Provides backward compatibility with firmware v1.5.3 and older and will be automatically set when
  the detected firmware version is v1.5.3 or older.

Configuration variables:
************************

> - **operating_mode** (*Required*): Provsions the operating mode select component.

Reviewer: ```suggestion
- **operating_mode** (*Required*): Provisions the operating mode select component.
```

 - **still_threshold** (Optional+Inclusive+Group2): Gate still value threashold level for motion energy detection on currently selected gate number. A value greater than that specified for the gate (distance) will trigger movement detection.
 - **move_threshold** (Optional+Inclusive+Group2): Gate move value threashold level for still energy detection on currently selected gate number. A value less than that specified for the gate (distance) will trigger still detection.

Button
-------

.. code-block:: yaml

    button:
      - platform: ld2420
        apply_config:
          name: (b) Apply Config
        factory_reset:
          name: (b) Factory Reset
        restart_module:
          name: (b) Restart Module
        revert_config:
          name: (b) Undo Edits

Four button components are available enabling configuration controls for editing, saving, restarting and factory reseting the LD2420 module.

Configuration variables:
************************

- **apply_config** (*Optional*): Saves both manual config tuning or the auto calibrate still and move threshold config settings.

- **restart_module** (*Optional*): Reboots the LD2420 modules.

> - **revert_config** (*Optional*): Undoes in progress edite prior to using the apply_config button.

Reviewer: ```suggestion
- **revert_config** (*Optional*): Undoes in-progress edits prior to their application via the ``apply_config`` button.
```

  for the duration of the **presence_time_window** setting.

Text Sensor
-----------

The ``ld2420`` text sensor provides version information for the :doc:`ld2420`.

.. code-block:: yaml

    text_sensor:
      - platform: ld2420
        fw_version:
          name: LD2420 Firmware

Configuration variables:
************************

- **fw_version** (*Optional*): Allows you to retrieve the :doc:`ld2420` firmware version.

Important Information:
**********************

Solid objects and noise outside the ``detection_gate_max`` and ``detection_gate_min`` may cause false
detections or result in abnormal gate thresholds. For example, a wall within the gate max range can result
in signal reflections. If your sensor reports unexpected detections, you should test it by placing it in a
completely open room with no moving objects.

Also, never place two or more sensors in a manner such that their detection fields overlap, as this will certainly
result in false detections.

Firmware update capability is available using the LD2420 tool provided by HLKTech.
You will need to email them and request the bin file.
At this time only firmware version v1.5.6 and up can be upgraded.

> **DO NOT** attemp to update older firmware versions. It will brick the module.

Reviewer: ```suggestion
**DO NOT** attempt to update to older firmware versions. It will brick the module.
```

          name : Moving Distance

    binary_sensor:
      - platform: ld2420
        has_target:
          name: Presence

    select:
      - platform: ld2420
        operating_mode:
          name: Operating Mode

    number:
      - platform: ld2420
        timeout:
          name: Detection Presence Timeout
        min_gate_distance:
          name: Detection Gate Minimum
        max_gate_distance:
          name: Detection Gate Maximum
        gate_select:
          name: Select Gate to Set
        still_threshold:
          name: Set Still Threshold Value
        move_threshold:
          name: Set Move Threshold Value

    button:
      - platform: ld2420
        apply_config:
          name: Apply Config
        factory_reset:
          name: Factory Reset
        restart_module:
          name: Restart Module
        revert_config:
          name: Undo Edits

Platform
--------

.. code-block:: yaml

    ld2420:

Configuration variables:
************************

**ld2420** (*Required*): Component platform.

Select
------

**Operating Modes:**

.. code-block:: yaml

    select:
      - platform: ld2420
        operating_mode:
          name: Operating Mode

> The dynamic UI provides 3 operational modes which are **Normal**, **Calibrate** and **Simple** using a select component.

Reviewer: ```suggestion
The dynamic UI provides three operational modes: **Normal**, **Calibrate** and **Simple**. The select component allows
you to choose the active mode.
```

> - **Normal**
>   This is the default mode. It sets the ld2420 in energy reporting mode and is availble with firmware v1.5.4 or greater.
>   When in this mode the module logs all gate energy levels continuouly in a local array making it available
>   for dynamic calibration functions.
> - **Calibrate**
>   When calibration mode is selected the module will collect the average and peak energy levels reporting them every 5 seconds.
>   The report is provided in the log console and for information only. In order to calibrate the gate configurations no
>   objects should be present in the target environment. The recording will establish the mmWave noise floor levels and
>   undesirable reflections or signal distortions peaks. This should be collected for at least 30 seconds. Longer runs can
>   catch more potential false signals when devices near the module are used such as Microwave ovens and other signal
>   emiting sources.
>   When the **apply_config** button is triggered in calibration mode the ld2420 will be configured with calculated values
>   based on its collected noise floor data and the module will be returned to Normal Mode. Selecting any other mode drops
>   the sessions collected noise data and no changes setting are staged. Once the **apply_config** action is performed it
>   will persist. The gate values can be manually adjusted or calibrated again or you can trigger the
>   factory reset button.
> - **Simple** Provides backward compatibility with firmware v1.5.3 and older and will be automatically set when
>   the detected firmware version is v1.5.3 or older.

Reviewer: ```suggestion
- **Normal**: This is the default mode. It sets the ld2420 in energy reporting mode and is availble with firmware
  v1.5.4 or greater. When in this mode the module logs all gate energy levels continuouly in a local array making it
  available for dynamic calibration functions.
- **Calibrate**: When calibration mode is selected the module will collect the average and peak energy levels,
  reporting them every five seconds. The report is provided in the log console for information only. In order to
  calibrate the gate configuration, no objects should be present in the target environment. The recording will
  establish the mmWave noise floor levels and undesirable reflections or signal distortions peaks. This should be
  collected for at least 30 seconds. Longer runs can catch more potential false signals when devices near the module
  are used such as microwave ovens and other sources of noise.

  When the **apply_config** button is pressed in calibration mode the LD2420 will be configured with calculated values
  based on its collected noise floor data and the module will return to Normal Mode. Selecting any other mode drops the
  session's collected noise data and no changed settings are staged. Once the **apply_config** action is performed, the
  computed values will then persist. The gate values can be manually adjusted or calibrated again or you can use the
  factory reset button to restore the module's default settings.
- **Simple**: Provides backward compatibility with firmware v1.5.3 and older and will be automatically set when the
  detected firmware version is v1.5.3 or older.
```

Configuration variables:
************************

- **operating_mode** (*Required*): Provisions the operating mode select component.

Number
------

.. code-block:: yaml

    number:
      - platform: ld2420
        presence_timeout:
          name: Detection Presence Timeout
        min_gate_distance:
          name: Detection Gate Minimum
        max_gate_distance:
          name: Detection Gate Maximum

        gate_select:
          name: Select Gate to Set
        still_threshold:
          name: Set Still Threshold Value
        move_threshold:
          name: Set Move Threshold Value

        gate_move_sensitivity:
          name: Move Calibration Sensitivity Factor
        gate_still_sensitivity:
          name: Still Calibration Sensitivity Factor

        gate_0:
          move_threshold:
            name: Gate 0 Move Threshold
          still_threshold:
            name: Gate 0 Still Threshold
        gate_1:
          move_threshold:
            name: Gate 1 Move Threshold
          still_threshold:
            name: Gate 1 Still Threshold

Configuration variables:
************************

>  - **timeout** (*Optional+Inclusive+Group1*): The time in seconds during which the occupied state (presence) will persist after presence is no longer detected. Any energy detection within the time window restarts the countdown from this value. Defaults to 120s.
>  - **min_gate_distance** (*Optional+Inclusive+Group1*): Minimum distance for move or still energy detection. Value between 0 and max_gate_distance - 1. Each increment equals 70 cm. Defaults to 1.
>  - **max_gate_distance** (*Optional+Inclusive+Group1*): Maximum gate for movement detection. Value from 1 to 15. Each gate detects movement and still energy at 70 cm increments. At the maximum value, the sensor’s range equates to 16 * 70 cm resulting in an approximately 12-meter range; however, accuracy at this range may be diminished. The recommended maximum value is 12 * 70 cm which results in approximately 9 meters.
>  - **gate_select** (*Optional+Inclusive+Group2*): Gate value to edit selection (Gate # from 0 to 15). When this variable is defined the component will run in compact UI mode. Gate thresholds are edited by selecting the gate number and then adjusting the move and still thresholds.
>  - **still_threshold** (*Optional+Inclusive+Group2*): Gate still value threshold level for motion energy detection on a currently selected gate number. A value greater than that specified for the gate (distance) will trigger movement detection.
>  - **move_threshold** (*Optional+Inclusive+Group2*): Gate move value threshold level for still energy detection on a currently selected gate number. A value less than that specified for the gate (distance) will trigger still detection.
>  - **gate_still_sensitivity** (*Optional*): Gate still threshold calibration sensitivity factor. Defaults as 0.5 with a range of 0-1, 0 = high and 1 = low.
>  - **gate_move_sensitivity** (*Optional*): Gate move threshold calibration sensitivity factor. Defaults as 0.5 with a range of 0-1, 0 = high and 1 = low
>  - **gate_n** (*Optional*): Provides individual gate threshold number inputs when gate_select is not defined. Range is gate_0 to gate_15. Each gate entry requires a still_threshold and move_threshold sub variable.
>   - **still_threshold** (*Required with gate_n entries*): Gate still value threshold level for motion energy detection on currently selected gate number. A value greater than that specified for the gate (distance) will trigger movement detection.
>   - **move_threshold** (*Required with gate_n entries*): Gate move value threshold level for still energy detection on currently selected gate number. A value less than that specified for the gate (distance) will trigger still detection.

Reviewer: ```suggestion
- **presence_timeout** (*Optional*): The time in seconds during which the occupied state (presence) will persist after
  presence is no longer detected. Any energy detection within the time window restarts the countdown from this value.
  Defaults to 120s. **Required** when ``min_gate_distance`` and ``max_gate_distance`` is specified.
- **min_gate_distance** (*Optional*): Minimum distance for move or still energy detection. Value between 0 and
  ``max_gate_distance`` - 1. Each increment equals 70 cm. Defaults to 1. **Required** when ``presence_timeout`` and
  ``max_gate_distance`` is specified.
- **max_gate_distance** (*Optional*): Maximum gate for movement detection. Value from 1 to 15. Each gate detects
  movement and still energy at 70 cm increments. At the maximum value, the sensor’s range equates to 16 * 70 cm
  resulting in an approximately 12-meter range; however, accuracy at this range may be diminished. The recommended
  maximum value is 12 * 70 cm which results in approximately 9 meters. **Required** when ``presence_timeout`` and
  ``min_gate_distance`` is specified.
- **gate_still_sensitivity** (*Optional*): Gate still threshold calibration sensitivity factor. Defaults to 0.5 with a
  range of 0-1, 0 = high and 1 = low.
- **gate_move_sensitivity** (*Optional*): Gate move threshold calibration sensitivity factor. Defaults to 0.5 with a
  range of 0-1, 0 = high and 1 = low.

- **gate_select** (*Optional*): Selection of gate value to edit (gate # from 0 to 15). Gate thresholds
  are edited by selecting the gate number and then adjusting the move and still thresholds. **Note that when this
  variable is defined, the component will run in compact UI mode and the following two configuration variables are then
  required:**

  - **still_threshold** (*Optional*): Gate still value threshold level for motion energy detection on a currently
    selected gate number. A value greater than that specified for the gate (distance) will trigger movement detection.
  - **move_threshold** (*Optional*): Gate move value threshold level for still energy detection on a currently selected
    gate number. A value less than that specified for the gate (distance) will trigger still detection.

- **gate_n** (*Optional*): Provides individual gate threshold number inputs. Range is ``gate_0`` to ``gate_15``.
  **May not be used with** ``gate_select`` (above). Each gate entry requires a ``still_threshold`` and ``move_threshold``:

    - **still_threshold** (*Required for each gate_n entry*): Gate still value threshold level for motion energy
      detection on currently selected gate number. A value greater than that specified for the gate (distance) will
      trigger movement detection.
    - **move_threshold** (*Required for each gate_n entry*): Gate move value threshold level for still energy
      detection on currently selected gate number. A value less than that specified for the gate (distance) will
      trigger still detection.

```

Button

> -------

Reviewer: ```suggestion
------
```

The ``ld2420`` text sensor provides version information for the :doc:`ld2420`.

.. code-block:: yaml

    text_sensor:
      - platform: ld2420
        fw_version:
          name: LD2420 Firmware

Configuration variables:
************************

- **fw_version** (*Optional*): Allows you to retrieve the :doc:`ld2420` firmware version.

Important Information:
**********************

Solid objects and noise outside the ``detection_gate_max`` and ``detection_gate_min`` may cause false
detections or result in abnormal gate thresholds. For example, a wall within the gate max range can result
in signal reflections. If your sensor reports unexpected detections, you should test it by placing it in a
completely open room with no moving objects.

Also, never place two or more sensors in a manner such that their detection fields overlap, as this will certainly
result in false detections.

Firmware update capability is available using the LD2420 tool provided by HLKTech.
You will need to email them and request the bin file.
At this time only firmware version v1.5.6 and up can be upgraded.
**DO NOT** attempt to update to older firmware versions. It will brick the module.

See Also
--------

> - Official Datasheet/Manuals are still in development for info email `sales@hlktech.com`

Reviewer: ```suggestion
- Official Datasheet/Manuals are still in development; for info email `sales@hlktech.com`.
```
